Automattic/wp-calypso
# Happychat: Handle socket disconnection

Closes #12183

This adds event handlers and reducers for handling the socket getting disconnected, to present a message to the user. Before this change, messages sent when disconnected were being sent nowhere with no information to users.

A disconnection notice was implemented in #11821, but the Redux store was not updated in response to disconnection so the notice wasn't shown.

 - [x] Handle disconnection event from socket.io
 - [x] Add reducer to change the status to `disconnected` in the Redux store
 - [x] Rewire the logic for deciding whether the Happychat window is shown. Currently, `isConnected` is used to determine whether to show the window. This is a problem because when the disconnection occurs, the window just disappears. We also need to make sure that the window and sidebar icon aren't shown before the first connection, however.
 - [x] Show timeline when disconnected

### How to test

1. Start up a local happychat service, point Calypso to it.
1. Start a chat
1. Kill the local happychat service.
1. The disconnected notice should appear in chat:
    ![screen shot 2017-03-24 at 12 59 21 pm](https://cloud.githubusercontent.com/assets/416133/24278800/be2ef038-1091-11e7-8026-a9fcc54f1c2f.png)
1. It will soon switch to:
    <img alt="screen shot 2017-03-24 at 12 56 07 pm" src="https://cloud.githubusercontent.com/assets/416133/24278727/517bc8e4-1091-11e7-9e48-af4a957c3275.png">
1. Restart the local happychat service
1. The notice should disappear, and you can continue chatting as usual
